modules/packages/OrderedSet/Treap.chpl

```diff
@@ -0,0 +1,872 @@
+
+      :arg e: The element to base
+      :type e: `eltType`
+
+      :returns: a tuple containing result
+      :rtype: `(bool, eltType)`
+    */
+    proc const successor(e: eltType): (bool, eltType) {
+      if !isDefaultInitializable(e) {
+        compilerError("successor is not avaliable on types that can't be \
+                      default-initialized, here: " + eltType: string);
+      }
+
+      _enter(); defer _leave();
+
+      var result: (bool, eltType);
+      result[0] = false;
+      var baseNode = _find(_root, e);
+      if baseNode == nil {
+        return result;
+      }
+
+      var resultNode = _neighbour(baseNode, 1);
+      if resultNode != nil {
+        result = (true, resultNode!.element);
+      }
+      return result;
+    }
+
+    /* Implementation for the `kth` procedure, without acquring the lock */
+    pragma "no doc"
+    proc _kth(node: nodeType, in k: int): nodeType {
+      if node == nil then return nil;
+
+      /* The size of children
+         If nil then 0 ( which is the default value for int )
+      */
+      var childSize: [0..#2] int;
+
+      if node!.children[0] != nil {
+        childSize[0] = node!.children[0]!.size;
+      }
+      if node!.children[1] != nil {
+        childSize[1] = node!.children[1]!.size;
+      }
+
+      // k-th in left subtree
+      if k <= childSize[0] then return _kth(node!.children[0], k);
+      k -= childSize[0];
+
+      // k-th is the current node
+      if k == 1 then return node;
+      k -= 1;
+
+      // k-th is in the right subtree
+      if k <= childSize[1] then return _kth(node!.children[1], k);
+
+      // else k-th is out of range, return nil
+      return nil;
+    }
+
+    /*
+      Find the k-th element in the orderedSet. k starts from 1.
+      Returns if there is such one element.
+      If there is, store the result in `result`.
+
+      :arg k: To find k-th element
+      :type k: `int`
+
+      :arg result: The destination to store the result
+      :type result: `eltType`
+
+      :return: if there is such one element
+      :rtype: `bool`
+    */
+    proc const kth(k: int): (bool, eltType) {
+      if !isDefaultInitializable(e) {
+        compilerError("kth is not avaliable on types that can't be \
+                      default-initialized, here: " + eltType: string);
+      }
+
+      _enter(); defer _leave();
+
+      var resultNode = _kth(_root, k);
+      if resultNode == nil {
+        var defaultValue: eltType;
+        return (false, defaultValue);
+      }
+      return (true, resultNode!.element);
+    }
+
+    /*
+      Returns the minimal element in the tree
+    */
+    pragma "no doc"
+    proc _first() {
+      var node = _root;
+      while node != nil && node!.children[0] != nil {
+        node = node!.children[0];
+      }
+      return node;
+    }
+
+    /*
+      Iterate over the elements of this orderedSet. Yields constant references
+      that cannot be modified.
```

Comment:
Should we say something about the yield order?

Reply:
Also you are using `orderedSet` but don't you mean `treap`? This happens other times in the doc as well.

Reply:
> Also you are using `orderedSet` but don't you mean `treap`? This happens other times in the doc as well.

Actually, I'm considering moving documentation to the OrderedSet module page. And here can use the conclusion from the ongoing discussion about how List should arrange the docuementation. Maybe at a latter time when we come to a conclusion on that.

Reply:
> Should we say something about the yield order?

I added more detail in OrderedSet's module documentation. It's kind of like the previous situation of Heap. Let me know if you have any doubts about this.

Reply:
I've moved all docs to OrderedSet module and put a "no doc" on Treap.chpl.